From 1084331510fd21b82cc2a4cc66e11d23e7610fbe Mon Sep 17 00:00:00 2001
From: Wenyou Yang <wenyou.yang@atmel.com>
Date: Fri, 30 Jan 2015 11:05:49 +0800
Subject: [PATCH] driver: macb: move the phys_enter_power_down() to the driver
 file

To decrease the the duplicated code, move int phys_enter_power_down(void)
from the board files to the driver file macb.c

The board implements the at91_eth0/1_hw_init() interface.

Signed-off-by: Wenyou Yang <wenyou.yang@atmel.com>
---
 board/sama5d3_xplained/sama5d3_xplained.c | 52 ++++++++++---------------------
 board/sama5d4_xplained/sama5d4_xplained.c | 30 ++++++------------
 board/sama5d4ek/sama5d4ek.c               | 52 ++++++++++---------------------
 driver/macb.c                             | 46 +++++++++++++++++++++++++--
 include/board.h                           |  5 ++-
 include/macb.h                            | 11 ++-----
 6 files changed, 91 insertions(+), 105 deletions(-)

diff --git a/board/sama5d3_xplained/sama5d3_xplained.c b/board/sama5d3_xplained/sama5d3_xplained.c
index e2e6360..e8d790a 100644
--- a/board/sama5d3_xplained/sama5d3_xplained.c
+++ b/board/sama5d3_xplained/sama5d3_xplained.c
@@ -44,7 +44,6 @@
 #include "arch/at91_pio.h"
 #include "arch/at91_ddrsdrc.h"
 #include "sama5d3_xplained.h"
-#include "macb.h"
 #include "twi.h"
 #include "act8865.h"
 
@@ -235,8 +234,10 @@ static void at91_special_pio_output_low(void)
 #endif
 
 #if defined(CONFIG_MAC0_PHY)
-static void gmac_hw_init(void)
+unsigned int at91_eth0_hw_init(void)
 {
+	unsigned int base_addr = AT91C_BASE_GMAC;
+
 	const struct pio_desc macb_pins[] = {
 		{"GMDC",	AT91C_PIN_PB(16), 0, PIO_DEFAULT, PIO_PERIPH_A},
 		{"GMDIO",	AT91C_PIN_PB(17), 0, PIO_DEFAULT, PIO_PERIPH_A},
@@ -245,12 +246,18 @@ static void gmac_hw_init(void)
 
 	pio_configure(macb_pins);
 	pmc_enable_periph_clock(AT91C_ID_PIOB);
+
+	pmc_enable_periph_clock(AT91C_ID_GMAC);
+
+	return base_addr;
 }
 #endif
 
 #if defined(CONFIG_MAC1_PHY)
-static void emac_hw_init(void)
+unsigned int at91_eth1_hw_init(void)
 {
+	unsigned int base_addr = AT91C_BASE_EMAC;
+
 	const struct pio_desc macb_pins[] = {
 		{"EMDC",	AT91C_PIN_PC(8), 0, PIO_DEFAULT, PIO_PERIPH_A},
 		{"EMDIO",	AT91C_PIN_PC(9), 0, PIO_DEFAULT, PIO_PERIPH_A},
@@ -259,49 +266,22 @@ static void emac_hw_init(void)
 
 	pio_configure(macb_pins);
 	pmc_enable_periph_clock(AT91C_ID_PIOC);
+
+	pmc_enable_periph_clock(AT91C_ID_EMAC);
+
+	return base_addr;
 }
 #endif
 
-#ifdef CONFIG_MACB
-int phys_enter_power_down(void)
+#if defined(CONFIG_MACB)
+void at91_disable_mac_clock(void)
 {
-	struct mii_bus macb_mii_bus;
-
 #if defined(CONFIG_MAC0_PHY)
-	gmac_hw_init();
-
-	macb_mii_bus.name = "GMAC KSZ9011RNI";
-	macb_mii_bus.reg_base = (void *)AT91C_BASE_GMAC;
-	macb_mii_bus.phy_addr = 1;
-
-	pmc_enable_periph_clock(AT91C_ID_GMAC);
-
-	if (phy_power_down_mode(&macb_mii_bus)) {
-		dbg_loud("%s: Failed to enter power down mode\n",
-						macb_mii_bus.name);
-	}
-
 	pmc_disable_periph_clock(AT91C_ID_GMAC);
 #endif
-
 #if defined(CONFIG_MAC1_PHY)
-	emac_hw_init();
-
-	macb_mii_bus.name = "EMAC KSZ8081RNB";
-	macb_mii_bus.reg_base = (void *)AT91C_BASE_EMAC;
-	macb_mii_bus.phy_addr = 1;
-
-	pmc_enable_periph_clock(AT91C_ID_EMAC);
-
-	if (phy_power_down_mode(&macb_mii_bus)) {
-		dbg_loud("%s: Failed to enter power down mode\n",
-						macb_mii_bus.name);
-	}
-
 	pmc_disable_periph_clock(AT91C_ID_EMAC);
 #endif
-
-	return 0;
 }
 #endif
 
diff --git a/board/sama5d4_xplained/sama5d4_xplained.c b/board/sama5d4_xplained/sama5d4_xplained.c
index 8b2964b..d8b9ece 100644
--- a/board/sama5d4_xplained/sama5d4_xplained.c
+++ b/board/sama5d4_xplained/sama5d4_xplained.c
@@ -49,7 +49,6 @@
 #include "matrix.h"
 #include "twi.h"
 #include "act8865.h"
-#include "macb.h"
 
 #if defined(CONFIG_REDIRECT_ALL_INTS_AIC)
 static void redirect_interrupts_to_aic(void)
@@ -572,8 +571,10 @@ static void SiI9022_hw_reset(void)
 }
 
 #if defined(CONFIG_MAC0_PHY)
-static void gmac0_hw_init(void)
+unsigned int at91_eth0_hw_init(void)
 {
+	unsigned int base_addr = AT91C_BASE_GMAC;
+
 	const struct pio_desc macb_pins[] = {
 		{"G0_MDC",	AT91C_PIN_PB(16), 0, PIO_DEFAULT, PIO_PERIPH_A},
 		{"G0_MDIO",	AT91C_PIN_PB(17), 0, PIO_DEFAULT, PIO_PERIPH_A},
@@ -582,32 +583,19 @@ static void gmac0_hw_init(void)
 
 	pio_configure(macb_pins);
 	pmc_enable_periph_clock(AT91C_ID_PIOB);
+
+	pmc_enable_periph_clock(AT91C_ID_GMAC);
+
+	return base_addr;
 }
 #endif
 
-#ifdef CONFIG_MACB
-int phys_enter_power_down(void)
+#if defined(CONFIG_MACB)
+void at91_disable_mac_clock(void)
 {
-	struct mii_bus macb_mii_bus;
-
 #if defined(CONFIG_MAC0_PHY)
-	gmac0_hw_init();
-
-	macb_mii_bus.name = "GMAC0 KSZ8081RNB";
-	macb_mii_bus.reg_base = (void *)AT91C_BASE_GMAC;
-	macb_mii_bus.phy_addr = 1;
-
-	pmc_enable_periph_clock(AT91C_ID_GMAC);
-
-	if (phy_power_down_mode(&macb_mii_bus)) {
-		dbg_loud("%s: Failed to enter power down mode\n",
-						macb_mii_bus.name);
-	}
-
 	pmc_disable_periph_clock(AT91C_ID_GMAC);
 #endif
-
-	return 0;
 }
 #endif
 
diff --git a/board/sama5d4ek/sama5d4ek.c b/board/sama5d4ek/sama5d4ek.c
index 7e8230b..868a1c9 100644
--- a/board/sama5d4ek/sama5d4ek.c
+++ b/board/sama5d4ek/sama5d4ek.c
@@ -50,7 +50,6 @@
 #include "matrix.h"
 #include "twi.h"
 #include "act8865.h"
-#include "macb.h"
 
 #if defined(CONFIG_REDIRECT_ALL_INTS_AIC)
 static void redirect_interrupts_to_aic(void)
@@ -584,8 +583,10 @@ static void SiI9022_hw_reset(void)
 }
 
 #if defined(CONFIG_MAC0_PHY)
-static void gmac0_hw_init(void)
+unsigned int at91_eth0_hw_init(void)
 {
+	unsigned int base_addr = AT91C_BASE_GMAC;
+
 	const struct pio_desc macb_pins[] = {
 		{"G0_MDC",	AT91C_PIN_PB(16), 0, PIO_DEFAULT, PIO_PERIPH_A},
 		{"G0_MDIO",	AT91C_PIN_PB(17), 0, PIO_DEFAULT, PIO_PERIPH_A},
@@ -594,12 +595,18 @@ static void gmac0_hw_init(void)
 
 	pio_configure(macb_pins);
 	pmc_enable_periph_clock(AT91C_ID_PIOB);
+
+	pmc_enable_periph_clock(AT91C_ID_GMAC);
+
+	return base_addr;
 }
 #endif
 
 #if defined(CONFIG_MAC1_PHY)
-static void gmac1_hw_init(void)
+unsigned int at91_eth1_hw_init(void)
 {
+	unsigned int base_addr = AT91C_BASE_GMAC1;
+
 	const struct pio_desc macb_pins[] = {
 		{"G1_MDC",	AT91C_PIN_PA(22), 0, PIO_DEFAULT, PIO_PERIPH_B},
 		{"G1_MDIO",	AT91C_PIN_PA(23), 0, PIO_DEFAULT, PIO_PERIPH_B},
@@ -608,49 +615,22 @@ static void gmac1_hw_init(void)
 
 	pio_configure(macb_pins);
 	pmc_enable_periph_clock(AT91C_ID_PIOA);
+
+	pmc_enable_periph_clock(AT91C_ID_GMAC1);
+
+	return base_addr;
 }
 #endif
 
-#ifdef CONFIG_MACB
-int phys_enter_power_down(void)
+#if defined(CONFIG_MACB)
+void at91_disable_mac_clock(void)
 {
-	struct mii_bus macb_mii_bus;
-
 #if defined(CONFIG_MAC0_PHY)
-	gmac0_hw_init();
-
-	macb_mii_bus.name = "GMAC0 KSZ8081RNB";
-	macb_mii_bus.reg_base = (void *)AT91C_BASE_GMAC;
-	macb_mii_bus.phy_addr = 1;
-
-	pmc_enable_periph_clock(AT91C_ID_GMAC);
-
-	if (phy_power_down_mode(&macb_mii_bus)) {
-		dbg_loud("%s: Failed to enter power down mode\n",
-						macb_mii_bus.name);
-	}
-
 	pmc_disable_periph_clock(AT91C_ID_GMAC);
 #endif
-
 #if defined(CONFIG_MAC1_PHY)
-	gmac1_hw_init();
-
-	macb_mii_bus.name = "GMAC1 KSZ8081RNB";
-	macb_mii_bus.reg_base = (void *)AT91C_BASE_GMAC1;
-	macb_mii_bus.phy_addr = 1;
-
-	pmc_enable_periph_clock(AT91C_ID_GMAC1);
-
-	if (phy_power_down_mode(&macb_mii_bus)) {
-		dbg_loud("%s: Failed to enter power down mode\n",
-						macb_mii_bus.name);
-	}
-
 	pmc_disable_periph_clock(AT91C_ID_GMAC1);
 #endif
-
-	return 0;
 }
 #endif
 
diff --git a/driver/macb.c b/driver/macb.c
index 1204e04..d5a1055 100644
--- a/driver/macb.c
+++ b/driver/macb.c
@@ -2,7 +2,7 @@
  *         ATMEL Microcontroller Software Support
  * ----------------------------------------------------------------------------
  * Copyright (c) 2014, Atmel Corporation
-
+ *
  * All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
@@ -132,6 +132,12 @@
 
 #define PHY_ID_NUMBER		(0x0022)
 
+struct mii_bus {
+	const char *name;
+	void *reg_base;
+	unsigned int phy_addr;
+};
+
 static inline unsigned int macb_read(void *base, unsigned int offset)
 {
 	return readl(base + offset);
@@ -344,7 +350,7 @@ static int phy_power_down(struct mii_bus *bus)
 	return 0;
 }
 
-int phy_power_down_mode(struct mii_bus *bus)
+static int phy_power_down_mode(struct mii_bus *bus)
 {
 	unsigned int phy_addr;
 	int ret = 0;
@@ -380,3 +386,39 @@ error:
 	macb_enable_managementport(bus, 0);
 	return ret;
 }
+
+int phys_enter_power_down(void)
+{
+	struct mii_bus macb_mii_bus;
+	unsigned int base_addr;
+
+#if defined(CONFIG_MAC0_PHY)
+	base_addr = at91_eth0_hw_init();
+
+	macb_mii_bus.name = "ETH0 PHY";
+	macb_mii_bus.reg_base = (void *)base_addr;
+	macb_mii_bus.phy_addr = 1;
+
+	if (phy_power_down_mode(&macb_mii_bus)) {
+		dbg_loud("%s: Failed to enter power down mode\n",
+						macb_mii_bus.name);
+	}
+#endif
+
+#if defined(CONFIG_MAC1_PHY)
+	base_addr = at91_eth1_hw_init();
+
+	macb_mii_bus.name = "ETH1 PHY";
+	macb_mii_bus.reg_base = (void *)base_addr;
+	macb_mii_bus.phy_addr = 1;
+
+	if (phy_power_down_mode(&macb_mii_bus)) {
+		dbg_loud("%s: Failed to enter power down mode\n",
+						macb_mii_bus.name);
+	}
+#endif
+
+	at91_disable_mac_clock();
+
+	return 0;
+}
diff --git a/include/board.h b/include/board.h
index 9ed583c..c0362a3 100644
--- a/include/board.h
+++ b/include/board.h
@@ -103,7 +103,10 @@ extern unsigned int at91_twi1_hw_init(void);
 extern unsigned int at91_twi2_hw_init(void);
 extern unsigned int at91_twi3_hw_init(void);
 
-extern int phys_enter_power_down(void);
+extern unsigned int at91_eth0_hw_init(void);
+extern unsigned int at91_eth1_hw_init(void);
+extern void at91_disable_mac_clock(void);
+
 extern void at91_disable_smd_clock(void);
 
 #endif /* #ifndef __BOARD_H__ */
diff --git a/include/macb.h b/include/macb.h
index 8934529..1ef00b9 100644
--- a/include/macb.h
+++ b/include/macb.h
@@ -2,7 +2,7 @@
  *         ATMEL Microcontroller Software Support
  * ----------------------------------------------------------------------------
  * Copyright (c) 2014, Atmel Corporation
-
+ *
  * All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
@@ -25,16 +25,9 @@
  * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
  * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
-
 #ifndef __MACB_H__
 #define __MACB_H__
 
-struct mii_bus {
-	const char *name;
-	void *reg_base;
-	unsigned int phy_addr;
-};
-
-extern int phy_power_down_mode(struct mii_bus *bus);
+extern int phys_enter_power_down(void);
 
 #endif /* #ifndef __MACB_H__ */
-- 
1.9.1

